lib/common/xml.c

```diff
@@ -1317,19 +1317,45 @@ write_xml_file(const xmlNode *xml, const char *filename, gboolean compress)
 
 /*!
  * \internal
- * \brief Check whether a string's first two bytes represent a Unicode charcter
+ * \brief Get consecutive bytes encoding non-ASCII UTF-8 characters
  *
  * \param[in] text  String to check
  *
- * \return \c true if the first two bytes of \p text represent a Unicode
- *         character, or \c false otherwise
+ * \return Number of non-ASCII UTF-8 bytes at the beginning of \p text
  */
-static inline bool
-is_unicode(const char *text)
+static size_t
+utf8_bytes(const char *text)
 {
-    // @TODO Is this a valid and complete test for multi-byte Unicode chars?
-    return !pcmk__str_empty(text)
-           && ((text[0] & 0x80) != 0) && ((text[1] & 0x80) != 0);
+    size_t utf8_bytes = 0;
+    size_t utf8_bits = 0;
+    size_t c_bytes = 0;
+
+    if (text == NULL) {
+        return 0;
+    }
+
+    /* UTF-8 uses 8-bit bytes.
+     * * 1-byte UTF-8 characters are the same as ASCII, and each begins with a
+     *   '0' bit.
+     * * In multi-byte UTF-8 characters, each byte begins with a '1' bit.
+     */
+    for (uint8_t *byte_8 = (uint8_t *) text; (*byte_8 & 0x80) != 0; byte_8++) {
```

Comment:
I revamped this and am about to push a new version. I realized that in rare cases where `CHAR_BIT != 0` and we find one UTF-8 character, the next character we check might not be at a C byte (char) boundary.

For example, if we have a 24-bit UTF-8 character and `CHAR_BIT` is 16, then we probably need to start the next check at bit 32.